File: docs/scripts/sync-mermaid-theme.mjs
#!/usr/bin/env node
/**
 * Ensures every ```mermaid block in docs/ starts with MERMAID_INIT.
 * Run from repo root: node docs/scripts/sync-mermaid-theme.mjs
 */
import { readFileSync, writeFileSync, readdirSync, statSync } from "fs";
import { join, relative } from "path";
import { MERMAID_INIT } from "../diagrams/mermaid-theme.mjs";

const docsRoot = join(import.meta.dirname, "..");
const skipFiles = new Set(["playbooks/mermaid.md"]);
const initLine = MERMAID_INIT;
const initRegex = /^%%\{init:[\s\S]*?\}%%[ \t]*(?:\r?\n[ \t]*)*/;

function walk(dir, files = []) {
  for (const name of readdirSync(dir)) {
    if (name.startsWith(".")) continue;
    const p = join(dir, name);
    const st = statSync(p);
    if (st.isDirectory()) walk(p, files);
    else if (name.endsWith(".md")) files.push(p);
  }
  return files;
}

function syncMermaidBlocks(content) {
  const fence = /```mermaid\n([\s\S]*?)```/g;
  let changed = false;
  const out = content.replace(fence, (_match, body) => {
    let next = body;
    if (initRegex.test(next.trimStart())) {
      next = next.replace(initRegex, `${initLine}\n`);
    } else {
      next = `${initLine}\n${next.replace(/^\n+/, "")}`;
    }
    if (next !== body) changed = true;
    return `\`\`\`mermaid\n${next}\`\`\``;
  });
  return { content: out, changed };
}

let updated = 0;
for (const file of walk(docsRoot)) {
  const normalizedRel = relative(docsRoot, file).replace(/\\/g, "/");
  if (skipFiles.has(normalizedRel)) continue;
  const raw = readFileSync(file, "utf8");
  const { content, changed } = syncMermaidBlocks(raw);
  if (changed) {
    writeFileSync(file, content, "utf8");
    console.log(`updated docs/${normalizedRel}`);
    updated++;
  }
}
console.log(updated ? `Done. ${updated} file(s).` : "All mermaid blocks already synced.");

File: docs/scripts/sync-mermaid-theme.py
#!/usr/bin/env python3
"""Ensure every Mermaid block in docs/ starts with the canonical init line."""

from __future__ import annotations

import re
import sys
from pathlib import Path

SCRIPT_DIR = Path(__file__).resolve().parent
DOCS_ROOT = SCRIPT_DIR.parent
REPO_ROOT = DOCS_ROOT.parent

sys.path.insert(0, str(REPO_ROOT))
from docs.diagrams.mermaid_theme import MERMAID_INIT  # noqa: E402

SKIP_FILES = {"playbooks/mermaid.md"}
INIT_RE = re.compile(r"%%\{init:[\s\S]*?\}%%[ \t]*(?:\r?\n[ \t]*)*")
FENCE_RE = re.compile(r"```mermaid\n([\s\S]*?)```")


def sync_mermaid_blocks(content: str) -> tuple[str, bool]:
    changed = False

    def replace(match: re.Match[str]) -> str:
        nonlocal changed
        body = match.group(1)
        stripped = body.lstrip()
        if INIT_RE.search(stripped):
            without_init = INIT_RE.sub("", stripped, count=1)
            next_body = f"{MERMAID_INIT}\n{without_init.lstrip(chr(10))}"
        else:
            next_body = f"{MERMAID_INIT}\n{body.lstrip(chr(10))}"

        if next_body != body:
            changed = True
        return f"```mermaid\n{next_body}```"

    return FENCE_RE.sub(replace, content), changed


def write_lf(path: Path, content: str) -> None:
    with path.open("w", encoding="utf-8", newline="\n") as handle:
        handle.write(content)


def main() -> int:
    updated = 0
    for path in DOCS_ROOT.rglob("*.md"):
        if any(part.startswith(".") for part in path.relative_to(DOCS_ROOT).parts):
            continue
        rel = path.relative_to(DOCS_ROOT).as_posix()
        if rel in SKIP_FILES:
            continue

        raw = path.read_text(encoding="utf-8")
        content, changed = sync_mermaid_blocks(raw)
        if changed:
            write_lf(path, content)
            print(f"updated docs/{rel}")
            updated += 1

    print(f"Done. {updated} file(s)." if updated else "All mermaid blocks already synced.")
    return 0


if __name__ == "__main__":
    raise SystemExit(main())
